http://gcc.gnu.org/ml/gcc/2006-01/msg01007.html
http://sourceware.org/ml/binutils/2006-02/msg00216.html

2006-02-17  Alan Modra  <amodra@bigpond.net.au>

	* elf32-ppc.c (allocate_dynrelocs): Tweak undef weak handling.
	* elf64-ppc.c (allocate_dynrelocs): Likewise.

===================================================================
RCS file: /cvs/src/src/bfd/elf32-ppc.c,v
retrieving revision 1.186
retrieving revision 1.187
diff -u -r1.186 -r1.187
--- src/bfd/elf32-ppc.c	2006/02/17 04:49:34	1.186
+++ src/bfd/elf32-ppc.c	2006/02/17 12:52:58	1.187
@@ -4546,7 +4546,8 @@
 
       /* Also discard relocs on undefined weak syms with non-default
 	 visibility.  */
-      if (h->root.type == bfd_link_hash_undefweak)
+      if (eh->dyn_relocs != NULL
+	  && h->root.type == bfd_link_hash_undefweak)
 	{
 	  if (ELF_ST_VISIBILITY (h->other) != STV_DEFAULT)
 	    eh->dyn_relocs = NULL;
===================================================================
RCS file: /cvs/src/src/bfd/elf64-ppc.c,v
retrieving revision 1.233
retrieving revision 1.234
diff -u -r1.233 -r1.234
--- src/bfd/elf64-ppc.c	2006/02/17 04:49:34	1.233
+++ src/bfd/elf64-ppc.c	2006/02/17 12:52:58	1.234
@@ -7582,7 +7582,8 @@
 
       /* Also discard relocs on undefined weak syms with non-default
 	 visibility.  */
-      if (h->root.type == bfd_link_hash_undefweak)
+      if (eh->dyn_relocs != NULL
+	  && h->root.type == bfd_link_hash_undefweak)
 	{
 	  if (ELF_ST_VISIBILITY (h->other) != STV_DEFAULT)
 	    eh->dyn_relocs = NULL;
